### PR TITLE
First try at using flatbuffers

### PR DESCRIPTION
**Pull request by olgavrou**
- Added the minimum flatbuffer files under 3rdparty
- ansible playbooks have been updated to build and install the flatc compiler
- The `frame.fbs` file is the schema which when compiled with the `flatc` compiler generates `frame_generated.h` which can then be used in the codebase
- Implemented `flatbufferwrapper.h` which we can use to use to further wrap the data vectors produced by the `GenericSerialisedWrapper` and use it for the final serialisation/deserialisation of the replicated and derived write sets
- updated docker image

### CODE REVIEW

**Comment by achamayou** [comment]
cgmanifest.json needs to be updated

File: 3rdparty/flatbuffers/include/flatbuffers/flatc.h [hunk 1]
```diff
@@ -0,0 +1,97 @@
+/*
```
Comment (jumaffre) [inline 3rdparty/flatbuffers/include/flatbuffers/flatc.h]: Do we need this file?

File: 3rdparty/flatbuffers/include/flatbuffers/grpc.h [hunk 2]
```diff
@@ -0,0 +1,330 @@
+/*
```
Comment (jumaffre) [inline 3rdparty/flatbuffers/include/flatbuffers/grpc.h]: Do we need this file?

File: cmake/common.cmake [hunk 3]
```diff
@@ -106,6 +111,8 @@ include_directories(
   ${CCF_DIR}/3rdparty
   ${CCF_DIR}/3rdparty/evercrypt-msr
   ${MSGPACK_INCLUDE_DIR}
+  ${FLATBUFFERS_INCLUDE_DIR}
+  ${CMAKE_CURRENT_BINARY_DIR}
```
Comment (jumaffre) [inline cmake/common.cmake]: Why is this required?

File: getting_started/setup_vm/roles/ccf_dependencies/tasks/install.yml [hunk 4]
```diff
@@ -59,4 +59,47 @@
   command: make install
   args:
     chdir: "{{ workspace }}/mbedtls-{{ mbedtls_dir }}/build"
-  become: true
\ No newline at end of file
+  become: true
+
+- name: Download FlatBuffers source
```
Comment (jumaffre) [inline getting_started/setup_vm/roles/ccf_dependencies/tasks/install.yml]: Would it be cleaner to have this in a different Ansible file?

File: cmake/common.cmake (hunk 3 above)
Comment (olgavrou) [reply cmake/common.cmake]: I am guessing you are referring to `CMAKE_CURRENT_BINARY_DIR`. It is because we are building `data_generated.h` during cmake and it will reside in the projects build directory. It is similar to `ccf_[u|v].h`

Comment (jumaffre) [reply cmake/common.cmake]: Oh I see, it's the location of the file generated by `flatc`. Should we keep the output of `flatc` under a specific sub-directory, e.g. `build/flatbuffers`?

File: src/kv/flatbufferwrapper.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include <data_generated.h>
```
Comment (jumaffre) [inline src/kv/flatbufferwrapper.h]: If we record the output of `flatc` not in `build/` (either in `build/flatbuffers/` or directly in `src/flatbuffers`), then it would be clearer where this file comes from.

Comment (olgavrou) [reply src/kv/flatbufferwrapper.h]: `src/flatbuffers` is suboptimal since we will then have to tell git to explicitly not commit those files and I don't think generated files should go under `src`. `build/flatbuffers` is an option to keep the `build` dir clean so we can think about that, I don't have a string opinion either way.

Comment (eddyashton) [reply src/kv/flatbufferwrapper.h]: I second generating in a build subfolder, preferably 2 layers deep so we can a) only add a subdir to the include directories, and b) clearly identify the generated headers.

eg - if this is `build/generated/flatbuffers/data_generated.h` then the cmake only adds `build/generated` to the include dirs, and this is `#include "flatbuffers/data_generated.h"`.

File: getting_started/setup_vm/roles/ccf_dependencies/tasks/install.yml (hunk 4 above)
Comment (achamayou) [reply getting_started/setup_vm/roles/ccf_dependencies/tasks/install.yml]: Only if we're going to use it somewhere else, I think we can burn that bridge when we come to it.

File: cmake/common.cmake [hunk 6]
```diff
@@ -4,6 +4,11 @@
 set(CMAKE_MODULE_PATH "${CCF_DIR}/cmake;${CMAKE_MODULE_PATH}")
 
 set(MSGPACK_INCLUDE_DIR ${CCF_DIR}/3rdparty/msgpack-c)
+set(FLATBUFFERS_INCLUDE_DIR ${CCF_DIR}/3rdparty/flatbuffers/include)
+
+execute_process(
+    COMMAND flatc --cpp ${CCF_DIR}/src/flatbuffers/data.fbs
```
Comment (achamayou) [inline cmake/common.cmake]: data is awfully generic, I suggest kv_frame instead. Alternatively src/kv/frame.fbs if we want to keep the idl files close to the source.

File: getting_started/setup_vm/roles/ccf_dependencies/vars/common.yml [hunk 7]
```diff
@@ -17,4 +17,9 @@ debs:
   - doxygen
 mbedtls_ver: "2.16.2"
 mbedtls_dir: "mbedtls-{{ mbedtls_ver }}"
-mbedtls_src: "{{ mbedtls_dir }}.tar.gz"
\ No newline at end of file
+mbedtls_src: "{{ mbedtls_dir }}.tar.gz"
+fb_ver: "1.11.0"
+flatbuffers_src: "flatbuffers-{{ fb_ver }}.tar.gz"
+flatbuffers_url: "https://github.com/google/flatbuffers/tarball/9e7e8cbe9f675123dd41b7c62868acad39188cae"
```
Comment (achamayou) [inline getting_started/setup_vm/roles/ccf_dependencies/vars/common.yml]: Why don't we pick up a release tarball?

File: src/kv/flatbufferwrapper.h (hunk 5 above)
Comment (achamayou) [reply src/kv/flatbufferwrapper.h]: fwiw, I think the build directory is for scripts, not for people, that's what install is for. I'm fine with subdirs, but I'm also fine without.

File: src/kv/flatbufferwrapper.h [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    const data::Data* data;
+
+  public:
+    FlatbufferDeserialiser(const std::vector<uint8_t>& data_) :
+      data(data::GetData(data_.data()))
+    {}
+
+    std::vector<uint8_t> get_replicated()
+    {
+      return std::move(std::vector<uint8_t>(
```
Comment (achamayou) [inline src/kv/flatbufferwrapper.h]: We don't want to make a copy.

File: src/kv/flatbufferwrapper.h [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+      auto fb_derived = builder.CreateVector(derived);
+
+      data = data::CreateData(builder, fb_replicated, fb_derived);
+      builder.Finish(data);
+    }
+
+    std::vector<uint8_t> get_flatbuffer()
+    {
+      auto buf = builder.GetBufferPointer();
+      return std::move(std::vector<uint8_t>(buf, buf + builder.GetSize()));
```
Comment (achamayou) [inline src/kv/flatbufferwrapper.h]: Likewise, we don't want to make a copy.

File: 3rdparty/flatbuffers/include/flatbuffers/flatc.h (hunk 1 above)
Comment (achamayou) [reply 3rdparty/flatbuffers/include/flatbuffers/flatc.h]: If we do, then we can build those and we don't need the ansible stuff. Otherwise let's remove it?

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images
